src/test/java/org/medicmobile/webapp/mobile/AppUrlVerifierTest.java

```diff
@@ -0,0 +1,161 @@
+package org.medicmobile.webapp.mobile;
+
+import java.io.IOException;
+import java.net.URL;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+import org.robolectric.annotation.Config;
+import static org.medicmobile.webapp.mobile.R.string.errAppUrl_apiNotReady;
+import static org.medicmobile.webapp.mobile.R.string.errAppUrl_appNotFound;
+import static org.medicmobile.webapp.mobile.R.string.errAppUrl_serverNotFound;
+import static org.medicmobile.webapp.mobile.R.string.errInvalidUrl;
+import static org.junit.Assert.*;
+
+@RunWith(RobolectricTestRunner.class)
+@Config(sdk=28)
+public class AppUrlVerifierTest {
+
+	private AppUrlVerifier buildAppUrlVerifier(JSONObject jsonResponse) {
```

Comment:
When possible I like to avoid having to hack like this with `protected` methods.  Java does not make this easy, but I think in this case we could use Mockito to inject a mock SimpleJsonClient2 into the AppUrlVerifier via a second constructor instead of overwriting the protected methods.  The first thing we would have to do is make the json client a class-level variable in AppUrlVerifier (and get rid of `getJsonClient()`) that is set by the constructors:

```.java
	private final SimpleJsonClient2 jsonClient;

	AppUrlVerifier(SimpleJsonClient2 jsonClient) {
		this.jsonClient = jsonClient;
	}

	public AppUrlVerifier() {
		this(new SimpleJsonClient2());
	}
```

Then, this test code can be updated to look something like this:

```.java
	private AppUrlVerifier buildAppUrlVerifier(JSONObject jsonResponse) {
		try {
			SimpleJsonClient2 mockJsonClient = Mockito.mock(SimpleJsonClient2.class);
			Mockito.when(mockJsonClient.get((String) ArgumentMatchers.any())).thenReturn(jsonResponse);
			return new AppUrlVerifier(mockJsonClient);
		} catch (Exception e) {
			throw new RuntimeException(e);
		}
	}

	private AppUrlVerifier buildAppUrlVerifierWithException(Exception e) {
		try {
			SimpleJsonClient2 mockJsonClient = Mockito.mock(SimpleJsonClient2.class);
			Mockito.when(mockJsonClient.get((String) ArgumentMatchers.any())).thenThrow(e);
			return new AppUrlVerifier(mockJsonClient);
		}  catch (Exception exception) {
			throw new RuntimeException(exception);
		}
	}

	private AppUrlVerifier buildAppUrlVerifierOk() {
		try {
			return buildAppUrlVerifier(new JSONObject("{\"ready\":true,\"handler\":\"medic-api\"}"));
		} catch (JSONException e) {
			throw new RuntimeException(e);
		}
	}
```

Reply:
I like the idea of injecting the Json client (Spring style :wink: ). I don't like Mockito too much, I think it overrides behaviors that the language already offer (like overriding behavior and injecting objects), allowing developers to write code with anti-patterns that the other way is less possible, because that kind of code is less testable unless... mockito.

Anyway, we already have Mockito in the project and it makes the code more compact in many cases (much appreciated when you write in Java), so I'll use it... against my will :stuck_out_tongue:

Reply:
Honestly, Mockito always makes me cringe too. Every time I use it I just get to thinking "there has to be a better way".  But, I have never been able to find one (unless your production code is already structured to leverage some kind of dependency injection or service provider framework).  

In this case it feels like we are using Mockito the "good" way in that we are just using it to construct objects with mocked behavior that we can inject (and not doing any kind of deep manipulation on the implementation code...).